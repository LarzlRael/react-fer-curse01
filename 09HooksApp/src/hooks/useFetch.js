import { useEffect, useRef, useState } from "react"


export const useFetch = (url) => {

    const [state, setState] = useState({ data: null, loading: true, error: null });

    // using the useRef when the component is mounted
    const isMounted = useRef(true);
    useEffect(() => {
        return () => {
            isMounted.current = false;
        }
    }, []);

    useEffect(() => {

        setState({
            loading: true,
            error: null,
            data: null
        });

        fetch(url)
            .then(resp => resp.json())
            .then(data => {


                if (isMounted.current) {
                    // simulate 4 secods delay request
                    setState({
                        loading: false,
                        error: null,
                        data
                    })
                }
            }).catch(() => {
                setState({
                    data: null,
                    loading: false,
                    error: 'No se pudo cargar la info prro',
                });
            } )
            
    }, [url])

    return state;
}
