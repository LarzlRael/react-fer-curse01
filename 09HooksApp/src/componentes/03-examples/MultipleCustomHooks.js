import React from 'react'
import { useCounter } from '../../hooks/useCounter';
import { useFetch } from '../../hooks/useFetch'


const MultipleCustomHooks = () => {

    const { counter, increment } = useCounter(1);

    const { loading, data } = useFetch(`https://www.breakingbadapi.com/api/quotes/${counter}`);

    // negando 2 veces para convertirlo en boolean
    const { author, quote } = !!data && data[0];
    console.log(loading);
    return (
        <div>
            <h1>BreakingBad Qoutes</h1>
            <hr />

            {loading ?
                <div className="alert alert-info text-center">
                    ...loading
                </div>
                :
                <blockquote className="blockquote text-right">
                    <p className="mb-0">{quote}</p>
                    <footer className="blockquote-footer">{author}</footer>
                </blockquote>
            }

            <button
                className="btn btn-success"
                onClick={increment}
            >Siguiente Frase </button>

        </div>
    )
}

export default MultipleCustomHooks;
