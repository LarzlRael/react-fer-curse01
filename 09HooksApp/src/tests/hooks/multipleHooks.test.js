import React from 'react';
import { shallow } from 'enzyme';
import MultipleCustomHooks from '../../componentes/03-examples/MultipleCustomHooks';
import { useFetch } from '../../hooks/useFetch';
import { useCounter } from '../../hooks/useCounter';

jest.mock('../../hooks/useFetch');
// jest.mock('../../hooks/useCounter');



describe('Testing in <MultipleCustomHooks />', () => {

    // useCounter.mockReturnValue({
    //     counter: 10,
    //     increment:() => {} 
    // })


    test('should display correctly', () => {

        useFetch.mockReturnValue({
            data: null,
            loading: true,
            error: null
        })

        const wrapper = shallow(<MultipleCustomHooks />)
        expect(wrapper).toMatchSnapshot();
    });

    test('should display the information', () => {

        const mensaje = 'hola mundo';
        useFetch.mockReturnValue({
            data: [{
                author: 'larz',
                quote: mensaje
            }],
            loading: false,
            error: null

        })

        const wrapper = shallow(<MultipleCustomHooks />)

        console.log(wrapper.html())
        expect(wrapper.find('.alert').exists()).toBe(false);
        expect(wrapper.find('.mb-0').text().trim()).toBe(mensaje);
        expect(wrapper.find('footer').text().trim()).toBe('larz');


    })


})


