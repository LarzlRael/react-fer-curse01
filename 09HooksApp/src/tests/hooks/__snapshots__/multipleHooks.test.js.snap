// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing in <MultipleCustomHooks /> should display correctly 1`] = `
<div>
  <h1>
    BreakingBad Qoutes
  </h1>
  <hr />
  <div
    className="alert alert-info text-center"
  >
    ...loading
  </div>
  <button
    className="btn btn-success"
    onClick={[Function]}
  >
    Siguiente Frase 
  </button>
</div>
`;
